File: android/app/src/main/java/info/bagen/rust/plaoc/microService/sys/nativeui/virtualKeyboard/VirtualKeyboardNMM.kt
package info.bagen.rust.plaoc.microService.sys.nativeui.virtualKeyboard

import info.bagen.rust.plaoc.microService.core.BootstrapContext
import info.bagen.rust.plaoc.microService.core.NativeMicroModule
import info.bagen.rust.plaoc.microService.helper.Mmid
import info.bagen.rust.plaoc.microService.sys.nativeui.NativeUiController
import info.bagen.rust.plaoc.microService.sys.nativeui.fromMultiWebView
import info.bagen.rust.plaoc.microService.sys.nativeui.helper.QueryHelper
import org.http4k.core.Method
import org.http4k.routing.bind
import org.http4k.routing.routes

class VirtualKeyboardNMM : NativeMicroModule("virtual-keyboard.nativeui.sys.dweb") {

    private fun getController(mmid: Mmid) =
        NativeUiController.fromMultiWebView(mmid).virtualKeyboard


    override suspend fun _bootstrap(bootstrapContext: BootstrapContext) {
        QueryHelper.init()
        apiRouting = routes(
            /** 获取状态 */
            "/getState" bind Method.GET to defineHandler { _, ipc ->
                return@defineHandler getController(ipc.remote.mmid)
            },
            /** 获取状态 */
            "/setState" bind Method.GET to defineHandler { request, ipc ->
                val controller = getController(ipc.remote.mmid)
                QueryHelper.overlay(request)?.also { controller.overlayState.value = it }
                QueryHelper.visible(request)?.also { controller.visibleState.value = it }
                return@defineHandler null
            },
            /**
             * 开始数据订阅
             */
            "/startObserve" bind Method.GET to defineHandler { _, ipc ->
                return@defineHandler getController(ipc.remote.mmid).observer.startObserve(ipc)
            },
            /**
             * 开始数据订阅
             */
            "/stopObserve" bind Method.GET to defineHandler { _, ipc ->
                return@defineHandler getController(ipc.remote.mmid).observer.stopObserve(ipc)
            },
        )
    }

    override suspend fun _shutdown() {
    }
}

File: next/dweb-browser/src/Platforms/iOS/MicroModule/NativeUI/VirtualKeyboard/VirtualKeyboardNMM.cs

using DwebBrowser.MicroService.Sys.NativeUI;

namespace DwebBrowser.Platforms.iOS.MicroModule.NativeUI.VirtualKeyboard;

public class VirtualKeyboardNMM : NativeMicroModule
{
    static Debugger Console = new("VirtualKeyboardNMM");
    public VirtualKeyboardNMM() : base("virtual-keyboard.nativeui.sys.dweb")
    {
    }

    private Task<VirtualKeyboardController> _getControllerAsync(Mmid mmid) =>
        MainThread.InvokeOnMainThreadAsync(() => NativeUiController.FromMultiWebView(mmid).VirtualKeyboard);

    // TODO：数据订阅路由功能未完成
    protected override async Task _bootstrapAsync(IBootstrapContext bootstrapContext)
    {
        /// 获取状态
        HttpRouter.AddRoute(IpcMethod.Get, "/getState", async (_, ipc) =>
        {
            return await _getControllerAsync(ipc.Remote.Mmid);
        });

        /// 设置状态
        HttpRouter.AddRoute(IpcMethod.Get, "/setState", async (request, ipc) =>
        {
            var controller = await _getControllerAsync(ipc.Remote.Mmid);

            request.QueryBool("overlay")?.Also(it =>
            {
                controller.OverlayState.Set(it);
                controller.Observer.Get();
            });

            /// 通过虚拟键盘事件来变更显示隐藏
            //request.QueryBool("visible")?.Also(it =>
            //    controller.VisibleState.Set(it));

            return null;
        });

        /// 开始数据订阅
        HttpRouter.AddRoute(IpcMethod.Get, "/startObserve", async (_, ipc) =>
        {
            return (await _getControllerAsync(ipc.Remote.Mmid)).StateObserver.StartObserve(ipc);
        });

        /// 停止数据订阅
        HttpRouter.AddRoute(IpcMethod.Get, "/stopObserve", async (_, ipc) =>
        {
            return (await _getControllerAsync(ipc.Remote.Mmid)).StateObserver.StopObserve(ipc);
        });
    }
}

